### PR TITLE
Fixes to while loops

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Adds tests for the while loop patterns used in codes and fixes the toolchain to pass those tests.

### CODE REVIEW

**Comment by jdahm** [comment]
The test in `test_code_generation.py` should probably be a stencil test suite, but I'm not sure how important the specific probability distributions are to the test, nor how to easily implement that logic in the test suites (uses hypothesis, which I'm not as familiar with).

File: tests/test_integration/test_code_generation.py [hunk 1]
```diff
@@ -382,23 +392,61 @@ def stencil_ijk(
     "backend", [backend for backend in ALL_BACKENDS if backend.values[0] != "gtc:dace"]
 )
 def test_variable_offsets_and_while_loop(backend):
+    nz = 100
```
[the remark this answers is not in the document]
Comment (jdahm) [reply tests/test_integration/test_code_generation.py:395]: I'm not completely happy with the test either. I agree with you that it's more than should be in a test suite, but I'm not sure how to simplify it.

This is a very specific situation that @leuty pointed out, where `K` indices go out of bounds for certain `IJ` points and not others. The test requires constants like `nz` to build `indK` and reproduce the behavior. It also draws from specific uniform and exponential probability distributions for a couple of the field inputs.

I tried making this a test suite (with python validation), but the problem is that `indK` is a constant array, and therefore doesn't fit the hypothesis strategy framework. I hacked in a way to make constant arrays, but I couldn't get it to work.

File: src/gtc/numpy/npir_codegen.py [hunk 2]
```diff
@@ -72,6 +72,8 @@ def shim_key(self, key):
                         new_args.append(idx + offset)
             if not isinstance(new_args[2], (numbers.Integral, slice)):
                 assert isinstance(new_args[0], slice) and isinstance(new_args[1], slice)
+                if np.max(new_args[2]) >= self.field_view.shape[2] or np.min(new_args[2]) < 0:
```
Comment (havogt) [inline src/gtc/numpy/npir_codegen.py:75]: I think that should be a function with a descriptive name, or at least a comment that describes this special case.

**Comment by jdahm** [comment]
bors try

**Pull request merged**